### DISCUSSION TITLE
Question about Objects/MyObject/MyVariable from example

### DISCUSSION THREAD

**Discussion by GronbergInfo** [FreeOpcUa/opcua-asyncio#966] [General]
Hi have this on a plc with opc ua enabled.

![example ocua](https://user-images.githubusercontent.com/12295022/178931309-6b3e6843-9ab0-482f-922b-53892502f2e2.png)

Would Objects be "root", MyObject = "Python" and MyVariable = "ManualEnabled" ?

When i run from the example:
        _logger.info("Children of root are: %r", await client.nodes.root.get_children())
The result i get is: 
    INFO:asyncua:Children of root are: [Node(NodeId(Identifier=85, NamespaceIndex=0, NodeIdType=<NodeIdType.TwoByte: 0>)), Node(NodeId(Identifier=86, NamespaceIndex=0, NodeIdType=<NodeIdType.TwoByte: 0>)), Node(NodeId(Identifier=87, NamespaceIndex=0, NodeIdType=<NodeIdType.TwoByte: 0>))]

**Comment by AndreasHeine** [comment]
that looks strange! you usually only have one root node...

what did you expect? kind of not understand your question!?

**Reply by GronbergInfo** [reply]
> that looks strange! you usually only have one root node...
> 
> what did you expect? kind of not understand your question!?

Hi. 
The "Python" is a folder in the plc/opc ua, where i wish to sort in my variables that i want to expose.
This is how it looks in the PLC setup.
![opc ua address space](https://user-images.githubusercontent.com/12295022/178935285-1f48d429-c8c0-4aed-9034-97edb9c7840e.png)

**Reply by AndreasHeine** [reply]
ok! how can i help?

**Reply by GronbergInfo** [reply]
> ok! how can i help?

I am trying to figure out how to access these variables in the python script i am making.

inthe example its written like this:
```
       # Now getting a variable node using its browse path
        myvar = await client.nodes.root.get_child(["0:Objects", "2:MyObject", "2:MyVariable"])
        obj = await client.nodes.root.get_child(["0:Objects", "2:MyObject"])
```
And im not sure which and what to change to access my current single variable.

**Reply by AndreasHeine** [reply]
get_child makes a BrowseRequest for HierarchicalReferences of the Node (client.nodes.root) you are calling it on. It returns a Node-Class-Instance if the BrowsePath (["0:Objects", "2:MyObject", "2:MyVariable"]) is valid in the OPC UA Servers Addressspace! However you can use different approaches "get_children()" returns all Nodes with HierarchicalReferences to that Node. And if you want to access the value you can just simply use the methods of the Node-Class e.g. "read_value()" (https://github.com/FreeOpcUa/opcua-asyncio/blob/db9e81d9a79efacfcda15f20140ca0ecd57fd505/asyncua/common/node.py#L39)

here is a small tutorial https://github.com/AndreasHeine/opcua-tutorial its not officially part of this repo its still "WIP" maybe some day ^^

**Reply by GronbergInfo** [reply]
> # Now getting a variable node using its browse path
>         myvar = await client.nodes.root.get_child(["0:Objects", "2:MyObject", "2:MyVariable"])

Thanks, after some digging around and reading "all over" the place, i managed to get this:

```
async def main():
    url = "opc.tcp://192.168.8.81:48020/"
    async with Client(url=url) as client:
        var = client.get_node("ns=4;i=2002")
        print(" --- myvar ---")
        myvar = client.get_node("ns=4;s=Root.Python.ManualEnabled")

        _logger.info("myvar is: %r", myvar)

    print("Value: ", await myvar.read_value()) # returns just the Value
        print("DataValue: ", await myvar.read_data_value()) # returns the DataValue (Value + StatusCode + Timestamps)
        print("BrowseName: ", await myvar.read_browse_name()) # returns the BrowseName ("0:ServiceLevel")
        print("DataType: ", await myvar.read_data_type()) # returns the DataType of the Value (i=3 -> Byte)
        print("DisplayName: ", await myvar.read_display_name()) # returns the DisplayName in form of LocalizedText
```

which in its turn i get this in the console:
```
 --- myvar ---
Value:  True
DataValue:  DataValue(Value=Variant(Value=True, VariantType=<VariantType.Boolean: 1>, Dimensions=None, is_array=False), StatusCode_=StatusCode(value=0), SourceTimestamp=datetime.datetime(2022, 7, 14, 9, 10, 16, 734289), ServerTimestamp=None, SourcePicoseconds=None, ServerPicoseconds=None)
BrowseName:  QualifiedName(NamespaceIndex=4, Name='ManualEnabled')
DataType:  NodeId(Identifier=1, NamespaceIndex=0, NodeIdType=<NodeIdType.TwoByte: 0>)
DisplayName:  LocalizedText(Locale=None, Text='ManualEnabled')
```

This is getting me plenty of the way, to read the values, i have tested and the Value changes from True to False  according to what i have in the PLC.

Thank you.

**Reply by AndreasHeine** [reply]
Your welcome!
